### PR TITLE
Support occupancy grid cells

### PR DESCRIPTION
**Pull request by SamerKhshiboun**
Tracked by [LRS-996]

### CODE REVIEW

File: realsense2_camera/src/base_realsense_node.cpp [hunk 1]
```diff
@@ -828,7 +839,47 @@ bool BaseRealSenseNode::shouldPublishCameraInfo(const stream_index_pair& sip)
...
+    auto cell_size = static_cast<float>(f.get_frame_metadata(RS2_FRAME_METADATA_OCCUPANCY_CELL_SIZE) / 100.0f); // convert to meters
+
+    // create GridCells msg and start filling it
+    nav_msgs::msg::GridCells msg;
+    msg.header.stamp = t;
+    msg.header.frame_id = FRAME_ID(OCCUPANCY);
+    msg.cell_width = cell_size;
+    msg.cell_height = cell_size;
+
+    for (auto i = 0; i < cols * rows; ++i)
```
Comment (Arun-Prasad-V) [inline realsense2_camera/src/base_realsense_node.cpp:862]: @SamerKhshiboun, just thinking out loud for future enhancement. Maybe it makes sense to parallelize this operation by using SSSE3 or GPU acceleration. your thoughts?

Comment (Nir-Az) [inline realsense2_camera/src/base_realsense_node.cpp:862]: Does this means rows and columns represent the cells and not the bytes?
meaning they are in `bit` units?
raws = 8 is 1 byte?

File: realsense2_camera/src/base_realsense_node.cpp [hunk 2]
```diff
@@ -828,7 +839,47 @@ bool BaseRealSenseNode::shouldPublishCameraInfo(const stream_index_pair& sip)
...
+    {
+        // check if current cell is occupied
+        // each byte include 8 bits <--> 8 cells
+        if ((frame_as_uint8_arr[i / 8U] & (1U << i % 8)) != 0) 
+        {
+            // find x,y,z positions of current index
+            geometry_msgs::msg::Point p3d;
+            uint32_t row = (i / cols);
+            uint32_t col = (i % cols);
+            p3d.x = (cell_size * static_cast<float>(rows)) - cell_size * (static_cast<float>(row) + 0.5f);
```
Comment (Nir-Az) [inline realsense2_camera/src/base_realsense_node.cpp:876]: I would expect x axis will use columns and not rows, can you explain this line?

File: realsense2_camera/src/base_realsense_node.cpp [hunk 3]
```diff
@@ -828,7 +839,47 @@ bool BaseRealSenseNode::shouldPublishCameraInfo(const stream_index_pair& sip)
...
+
+    for (auto i = 0; i < cols * rows; ++i)
+    {
+        // check if current cell is occupied
+        // each byte include 8 bits <--> 8 cells
+        if ((frame_as_uint8_arr[i / 8U] & (1U << i % 8)) != 0) 
+        {
+            // find x,y,z positions of current index
+            geometry_msgs::msg::Point p3d;
+            uint32_t row = (i / cols);
```
Comment (Nir-Az) [inline realsense2_camera/src/base_realsense_node.cpp:874]: raw is calculates from cols?
Why not from `RS2_FRAME_METADATA_OCCUPANCY_GRID_ROWS`?

File: realsense2_camera/src/base_realsense_node.cpp (hunk 2 above)
Comment (SamerKhshiboun) [reply realsense2_camera/src/base_realsense_node.cpp:876]: in ROS CS, X is forward, Y is left, and Z is up.
in AICV algo ROWS represents the height is and COLS represents the width.
image

File: realsense2_camera/src/base_realsense_node.cpp (hunk 1 above)
Comment (SamerKhshiboun) [reply realsense2_camera/src/base_realsense_node.cpp:862]: Yes, I think it worth checking if we can accelerate this.
If so, we can do this in a separated task.
Thanks for the advice !

Comment (SamerKhshiboun) [reply realsense2_camera/src/base_realsense_node.cpp:862]: rows = grid cell height, cols = grid cell width
added more documentation for this code

**Review by Nir-Az — APPROVED** [review]

**Pull request merged**